lib/data_provider.js

```diff
@@ -0,0 +1,128 @@
+/* Copyright 2021 F5 Networks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const fs = require('fs');
+
+const ResourceCache = require('./resource_cache').ResourceCache;
+
+/**
+ * DataProvider that fetches data from the file system
+ */
+class FsDataProvider {
+    /**
+     * @param {string} dataRootPath - a path to a directory containing data files
+     */
+    constructor(dataRootPath) {
+        this.data_path = dataRootPath;
+        this.cache = new ResourceCache(dataName => new Promise((resolve, reject) => {
+            fs.readFile(`${dataRootPath}/${dataName}.data`, (err, data) => {
+                if (err) return reject(err);
+                return resolve(data.toString('utf8'));
+            });
+        }));
+    }
+
+    /**
+     * Get the data file contents associated with the supplied key
+     *
+     * @param {string} key
+     * @returns {object}
+     */
+    fetch(key) {
+        return this.cache.fetch(key);
+    }
+
+    /**
+     * List all data files known to the provider
+     *
+     * @returns {string[]}
+     */
+    list() {
+        return new Promise((resolve, reject) => {
+            fs.readdir(this.data_path, (err, data) => {
+                if (err) return reject(err);
+
+                const list = data.filter(x => x.endsWith('.data'))
+                    .map(x => x.replace(/.data$/, ''));
+                return resolve(list);
+            });
+        });
+    }
+}
+
+/**
+ * DataProvider that fetches data from an atg-storage DataStore
+ */
+class DataStoreDataProvider {
```

Comment:
Would it be better to create a "base" class DataProvider and have some classes extend it?

```
class BaseDataProvider {
    constructor(dataStore, cache) {
      this.storage = dataStore;
       this.cache = cache;
  }

  fetch(key) {
   return this.cache.fetch(key)
 }
...
...
 }
```

Also, the name seems to be for generic purpose but it has a property/verbiage specifically for template sets. Maybe we can do something like below?

```
class DataStoreDataProvider {
    constructor(datastore, datastoreKey) {
        this.dsKey = datastoreKey;
        super(datastore, new ResourceCache(
               dataName => dataStore.hasItem(this.dsKey)
...
...

   }
}
```

Reply:
I have created a common `BaseDataProvider` class to address the first point. For the second point, all Providers currently operate on template sets.

Reply:
I have created a common BaseDataProvider class to address the first point. For the second point, all Providers currently operate on template sets.